src/compiler/file_uploader/index.ts

```diff
@@ -0,0 +1,187 @@
+        }
+    } catch (error) {
+        console.error(`Error reading directory: ${error}`);
+    }
+}
+
+async function uploadFile(
+    srcPath: Src,
+    destPath: Dest,
+    chunkSize: number,
+    actor: ActorSubclass
+) {
+    console.info(`uploadFile: Uploading ${srcPath} to ${destPath}`);
+    const uploadStartTime = process.hrtime.bigint();
+    const fileSize = (await stat(srcPath)).size;
+    for (let i = 0; i < fileSize; i += chunkSize) {
+        const fileStream = createReadStream(srcPath, {
+            start: i,
+            end: i + chunkSize - 1,
+            highWaterMark: chunkSize
+        });
+
+        let startIndex = i;
+        for await (const data of fileStream) {
+            await throttle();
+            console.info(
+                `uploadFile: ${srcPath} | ${bytesToHumanReadable(
+                    i + data.length
+                )} of ${bytesToHumanReadable(fileSize)}`
+            );
+            // Don't await here! Awaiting the agent will result in about a 4x increase in upload time.
+            // The above throttling is sufficient to manage the speed of uploads
+            actor
+                .upload_file_chunk(destPath, uploadStartTime, i, data, fileSize)
+                .catch((error) => {
+                    console.error(error);
+                });
+            startIndex += data.length;
+        }
+    }
+    console.info(`uploadFile: finished ${srcPath}`);
+}
+
+function bytesToHumanReadable(sizeInBytes: number): string {
+    const suffixes = ['B', 'KiB', 'MiB', 'GiB'];
+    // TODO it would be nice if we could figure out an algorithm as simple as
+    // TODO this that didn't involve a mutation
+    let size = sizeInBytes;
+
+    for (const suffix in suffixes) {
+        if (size < 1024.0) {
+            return `${size.toFixed(2)} ${suffixes[suffix]}`;
+        }
+        size /= 1024.0;
+    }
+
+    return `${size.toFixed(2)} ${suffixes[suffixes.length - 1]}`;
+}
+
+async function throttle() {
+    // We can only process about 4Mib per second. So if chunks are about
+    // 2 MiB or less then we can only send off two per second.
+    if (process.env.DFX_NETWORK === 'ic') {
+        await new Promise((resolve) => setTimeout(resolve, 2_000)); // Mainnet requires more throttling. We found 2_000 by trial and error
+    }
+    await new Promise((resolve) => setTimeout(resolve, 500)); // Should be 500 (ie 1 every 1/2 second or 2 every second)
+}
+
+async function getIdentity(
+    identityName: string
+): Promise<Secp256k1KeyIdentity> {
+    const identityPath = join(
+        homedir(),
+        '.config',
+        'dfx',
+        'identity',
+        identityName,
+        'identity.pem'
+    );
+    return Secp256k1KeyIdentity.fromPem(await readFile(identityPath, 'utf-8'));
+}
+
+async function createUploadFileChunkActor(
+    canisterId: string,
+    replicaWebServerPort: string,
+    identityName: string
+): Promise<ActorSubclass> {
+    const host =
```

Comment:
Just call getAgentHost, move the mainnet stuff into getAgentHost

Reply:
done